pkg/provider/provider.go

```diff
@@ -350,22 +403,84 @@ func (p *Provider) getKeyVaultObjectContent(ctx context.Context, kvClient *kv.Ba
 
 	switch kvObject.ObjectType {
 	case VaultObjectTypeSecret:
-		return getSecret(ctx, kvClient, vaultURL, kvObject)
+		return getSecret(ctx, kvClient, vaultURL, kvObject, objectVersion)
 	case VaultObjectTypeKey:
-		return getKey(ctx, kvClient, vaultURL, kvObject)
+		return getKey(ctx, kvClient, vaultURL, kvObject, objectVersion)
 	case VaultObjectTypeCertificate:
-		return getCertificate(ctx, kvClient, vaultURL, kvObject)
+		return getCertificate(ctx, kvClient, vaultURL, kvObject, objectVersion)
 	default:
 		err := errors.Errorf("Invalid vaultObjectTypes. Should be secret, key, or cert")
-		return "", "", wrapObjectTypeError(err, kvObject.ObjectType, kvObject.ObjectName, kvObject.ObjectVersion)
+		return "", "", wrapObjectTypeError(err, kvObject.ObjectType, kvObject.ObjectName, objectVersion)
 	}
 }
 
+func (p *Provider) getKeyVaultSecretVersions(ctx context.Context, kvClient *kv.BaseClient, kvObject KeyVaultObject, vaultURL string) (versions []string, err error) {
+	start := time.Now()
+	defer func() {
+		var errMsg string
+		if err != nil {
+			errMsg = err.Error()
+		}
+		p.reporter.ReportKeyvaultRequest(ctx, time.Since(start).Seconds(), kvObject.ObjectType, kvObject.ObjectName, errMsg)
+	}()
+
+	// We have to pull all of the secret versions for a secret because we can't trust
+	// the values returned by KV to be in any reliable order
+	kvVersionsList, err := kvClient.GetSecretVersions(ctx, vaultURL, kvObject.ObjectName, nil)
+	if err != nil {
+		return nil, wrapObjectTypeError(err, kvObject.ObjectType, kvObject.ObjectName, kvObject.ObjectVersion)
+	}
+
+	secretVersions := KeyVaultSecretList{}
+
+	for notDone := true; notDone; notDone = kvVersionsList.NotDone() {
+		for _, secret := range kvVersionsList.Values() {
+			objectVersion := getObjectVersion(*secret.ID)
+			created := date.UnixEpoch()
+
+			if secret.Attributes != nil {
+				created = time.Time(*secret.Attributes.Created)
+			}
+
+			if secret.Attributes.Enabled != nil && *secret.Attributes.Enabled {
+				secretVersions = append(secretVersions, KeyVaultSecret{
+					Version: objectVersion,
+					Created: created,
+				})
+			}
+		}
+
+		err = kvVersionsList.NextWithContext(ctx)
+		if err != nil {
+			return nil, wrapObjectTypeError(err, kvObject.ObjectType, kvObject.ObjectName, kvObject.ObjectVersion)
+		}
+	}
+
+	sort.Sort(secretVersions) // by creation date
+
+	// if we're being asked for the latest, then there's no need to skip any versions
+	var foundFirst = kvObject.ObjectVersion == "" || kvObject.ObjectVersion == "latest"
+
+	for _, secret := range secretVersions {
+		foundFirst = foundFirst || secret.Version == kvObject.ObjectVersion
+
+		if foundFirst {
+			versions = append(versions, secret.Version)
+		}
+
+		if len(versions) >= int(kvObject.ObjectVersionHistory) {
+			break
+		}
+	}
+
+	return versions, nil
+}
```

Comment:
Most of this after `GetSecretVersions` can be unit tested if split into smaller functions

Reply:
I've split things up a bit more and will add some additional unit tests